File: src/csharp/pulse/cdm/properties/SESegmentSigmoidal.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class SESegmentSigmoidal : SESegment
  {
    protected SEScalarPressure lowerCorner;
    protected SEScalarPressure upperCorner;
    protected SEScalarVolumePerPressure baselineCompliance;

    public SESegmentSigmoidal() : base()
    {
      lowerCorner = null;
      upperCorner = null;
      baselineCompliance = null;
    }

    public SESegmentSigmoidal(SESegment other)
    {
      Copy(other);
    }

    public void Copy(SESegmentSigmoidal other)
    {
      base.Copy(other);
      if (other.HasLowerCorner())
        GetLowerCorner().Set(other.GetLowerCorner());
      if (other.HasUpperCorner())
        GetUpperCorner().Set(other.GetUpperCorner());
      if (other.HasBaselineCompliance())
        GetBaselineCompliance().Set(other.GetBaselineCompliance());
    }

    public override void Clear()
    {
      base.Clear();
      if (lowerCorner != null)
        lowerCorner.Invalidate();
      if (upperCorner != null)
        upperCorner.Invalidate();
      if (baselineCompliance != null)
        baselineCompliance.Invalidate();
    }

    public override bool IsValid()
    {
      if (!base.IsValid())
        return false;
      return HasLowerCorner() && HasUpperCorner() &&
        HasBaselineCompliance();
    }

    public bool HasLowerCorner()
    {
      return lowerCorner == null ? false : lowerCorner.IsValid();
    }
    public SEScalarPressure GetLowerCorner()
    {
      if (lowerCorner == null)
        lowerCorner = new SEScalarPressure();
      return lowerCorner;
    }

    public bool HasUpperCorner()
    {
      return upperCorner == null ? false : upperCorner.IsValid();
    }
    public SEScalarPressure GetUpperCorner()
    {
      if (upperCorner == null)
        upperCorner = new SEScalarPressure();
      return upperCorner;
    }

    public bool HasBaselineCompliance()
    {
      return baselineCompliance == null ? false : baselineCompliance.IsValid();
    }
    public SEScalarVolumePerPressure GetBaselineCompliance()
    {
      if (baselineCompliance == null)
        baselineCompliance = new SEScalarVolumePerPressure();
      return baselineCompliance;
    }


    public override string ToString()
    {
      string str = base.ToString();
      str += "\n\tLowerCorner: " + (HasLowerCorner() ? GetLowerCorner().ToString() : "Not Provided");
      str += "\n\tUpperCorner: " + (HasUpperCorner() ? GetUpperCorner().ToString() : "Not Provided");
      str += "\n\tBaselineCompliance: " + (HasBaselineCompliance() ? GetBaselineCompliance().ToString() : "Not Provided");

      return str;
    }
  }
}

File: src/java/com/kitware/pulse/cdm/properties/SESegmentSigmoidal.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package com.kitware.pulse.cdm.properties;

import com.kitware.pulse.cdm.bind.Properties.SegmentSigmoidalData;

public class SESegmentSigmoidal extends SESegment
{
  protected SEScalarPressure          lowerCorner;
  protected SEScalarPressure          upperCorner;
  protected SEScalarVolumePerPressure baselineCompliance;

  public SESegmentSigmoidal()
  {
    super();
    lowerCorner = null;
    upperCorner = null;
    baselineCompliance = null;
  }

  public SESegmentSigmoidal(SESegment other)
  {
    copy(other);
  }

  public void copy(SESegmentSigmoidal other)
  {
    super.copy(other);
    if (other.hasLowerCorner())
      getLowerCorner().set(other.getLowerCorner());
    if (other.hasUpperCorner())
      getUpperCorner().set(other.getUpperCorner());
    if (other.hasBaselineCompliance())
      getBaselineCompliance().set(other.getBaselineCompliance());
  }

  public void clear()
  {
    super.clear();
    if (lowerCorner != null)
      lowerCorner.invalidate();
    if (upperCorner != null)
      upperCorner.invalidate();
    if (baselineCompliance != null)
      baselineCompliance.invalidate();
  }

  public static void load(SegmentSigmoidalData src, SESegmentSigmoidal dst)
  {
    dst.clear();
    if (src.hasBeginVolume())
      SEScalarVolume.load(src.getBeginVolume(), dst.getBeginVolume());
    if (src.hasEndVolume())
      SEScalarVolume.load(src.getEndVolume(), dst.getEndVolume());
    if (src.hasLowerCorner())
      SEScalarPressure.load(src.getLowerCorner(), dst.getLowerCorner());
    if (src.hasUpperCorner())
      SEScalarPressure.load(src.getUpperCorner(), dst.getUpperCorner());
    if (src.hasBaselineCompliance())
      SEScalarVolumePerPressure.load(src.getBaselineCompliance(), dst.getBaselineCompliance());
  }

  public static SegmentSigmoidalData unload(SESegmentSigmoidal src)
  {
    if (!src.isValid())
      return null;
    SegmentSigmoidalData.Builder dst = SegmentSigmoidalData.newBuilder();
    unload(src, dst);
    return dst.build();
  }

  protected static void unload(SESegmentSigmoidal src, SegmentSigmoidalData.Builder dst)
  {
    if (src.hasBeginVolume())
      dst.setBeginVolume(SEScalarVolume.unload(src.beginVolume));
    if (src.hasEndVolume())
      dst.setEndVolume(SEScalarVolume.unload(src.endVolume));
    if (src.hasLowerCorner())
      dst.setLowerCorner(SEScalarPressure.unload(src.lowerCorner));
    if (src.hasUpperCorner())
      dst.setUpperCorner(SEScalarPressure.unload(src.upperCorner));
    if (src.hasBaselineCompliance())
      dst.setBaselineCompliance(SEScalarVolumePerPressure.unload(src.baselineCompliance));
  }

  public boolean isValid()
  {
    if (!super.isValid())
      return false;
    return hasLowerCorner() && hasUpperCorner() && hasBaselineCompliance();
  }

  public boolean hasLowerCorner()
  {
    return lowerCorner == null ? false : lowerCorner.isValid();
  }

  public SEScalarPressure getLowerCorner()
  {
    if (lowerCorner == null)
      lowerCorner = new SEScalarPressure();
    return lowerCorner;
  }

  public boolean hasUpperCorner()
  {
    return upperCorner == null ? false : upperCorner.isValid();
  }

  public SEScalarPressure getUpperCorner()
  {
    if (upperCorner == null)
      upperCorner = new SEScalarPressure();
    return upperCorner;
  }

  public boolean hasBaselineCompliance()
  {
    return baselineCompliance == null ? false : baselineCompliance.isValid();
  }

  public SEScalarVolumePerPressure getBaselineCompliance()
  {
    if (baselineCompliance == null)
      baselineCompliance = new SEScalarVolumePerPressure();
    return baselineCompliance;
  }

  public String toString()
  {
    String str = super.toString();
    str += "\n\tLowerCorner: " + (hasLowerCorner() ? getLowerCorner().toString() : "Not Provided");
    str += "\n\tUpperCorner: " + (hasUpperCorner() ? getUpperCorner().toString() : "Not Provided");
    str += "\n\tBaselineCompliance: " + (hasBaselineCompliance() ? getBaselineCompliance().toString() : "Not Provided");

    return str;
  }
}
